src/client/components/pages/RoomAdmin/__tests__/RoomAdminTable.test.tsx

```diff
@@ -0,0 +1,117 @@
+import React from 'react';
+import {
+  strictEqual,
+} from 'assert';
+import {
+  wait,
+} from '@testing-library/react';
+import {
+  stub,
+  SinonStub,
+} from 'sinon';
+import { render } from 'test-utils';
+import { LocationAPI } from 'client/api';
+import {
+  adminRoomsResponse,
+  bauerRoomResponse,
+  error,
+  oxfordRoomResponse,
+  secRoomResponse,
+} from 'testData';
+import RoomAdmin from '../RoomAdmin';
+
+describe('Room Admin Table', function () {
+  let getStub: SinonStub;
+  describe('data fetching', function () {
+    context('when room data fetch succeeds', function () {
+      beforeEach(function () {
+        getStub = stub(LocationAPI, 'getAdminRooms');
+        getStub.resolves(adminRoomsResponse);
+      });
+      context('when there are room records', function () {
+        it('displays one table row per room', async function () {
+          const { getAllByRole } = render(
+            <RoomAdmin />
+          );
+          await wait(() => getAllByRole('row').length > 1);
+          const rows = getAllByRole('row');
+          // The "+1" takes the table headers into account for the number of rows
+          strictEqual(rows.length, adminRoomsResponse.length + 1);
+        });
+        it('displays the correct content in the table cells', async function () {
+          const { getAllByRole } = render(
+            <RoomAdmin />
+          );
+          await wait(() => getAllByRole('row').length > 1);
+          const rows = Array.from(getAllByRole('row')) as HTMLTableRowElement[];
+          const rowsContent = rows
+            .map(
+              (row) => (Array.from(row.cells).map((cell) => cell.textContent))
+            );
+          // An amalgamation of all room information
+          const roomInfo = [
+            secRoomResponse.name,
+            secRoomResponse.building.name,
+            secRoomResponse.building.campus.name,
+            secRoomResponse.capacity.toString(),
+            oxfordRoomResponse.name,
+            oxfordRoomResponse.building.name,
+            oxfordRoomResponse.building.campus.name,
+            oxfordRoomResponse.capacity.toString(),
+            bauerRoomResponse.name,
+            bauerRoomResponse.building.name,
+            bauerRoomResponse.building.campus.name,
+            bauerRoomResponse.capacity.toString(),
+          ];
+
+          for (let i = 1; i < rows.length; i++) {
+            rowsContent[i] = rowsContent[i].slice(0, -1);
+            rowsContent[i].forEach((data) => {
+              // Remove the Edit column, which contains no table data.
+              strictEqual(
+                roomInfo.includes(data),
+                true,
+                `${data} is not in the table as expected`
+              );
+            });
+          }
```

Comment:
So I was going to make a few suggestions here, but I've grouped them into one comment to go with the code snippet below.

- I don't think we need the calls to `Array.from()` here since `getAllByRole` already returns an array, and `row.cells` should already be an array.
- We probably want to be using react testing library as much as we can to do the work for us, rather than doing array slicing and dicing of DOM elements. As long as those fields exist _somewhere_ in the table row....the test should still pass.

```suggestion
          const rows = getAllByRole('row').slice(1) as HTMLTableRowElement[];
          
          [
            [
              secRoomResponse.name,
              secRoomResponse.building.name,
              secRoomResponse.building.campus.name,
              secRoomResponse.capacity.toString(),
            ],
            [
              oxfordRoomResponse.name,
              oxfordRoomResponse.building.name,
              oxfordRoomResponse.building.campus.name,
              oxfordRoomResponse.capacity.toString(),
            ],
            [
              bauerRoomResponse.name,
              bauerRoomResponse.building.name,
              bauerRoomResponse.building.campus.name,
              bauerRoomResponse.capacity.toString(),
            ],
          ].forEach((roomData, index) => {
            roomData.forEach((data) => {
              strictEqual(
                within(rows[index]).queryByText(data) !== null,
                true,
                `${data} is not in the table as expected`
              );
            });
          });
```

Reply:
Thanks for the suggestion Robert. That makes a lot of sense to try and take advantage of the testing library as much as we can. I applied the changes in commit [2081609](https://github.com/seas-computing/course-planner/pull/570/commits/2081609321cd532f08180160d4795acbfeec69ad).